airflow/providers/hashicorp/hooks/vault.py

```diff
@@ -0,0 +1,546 @@
+    :type auth_type: str
+    :param mount_point: The "path" the secret engine was mounted on. Default is "secret". Note that
+         this mount_point is not used for authentication if authentication is done via a
+         different engine.
+    :type mount_point: str
+    :param kv_engine_version: Select the version of the engine to run (``1`` or ``2``). Defaults to
+          version defined in connection or ``2`` if not defined in connection.
+    :type kv_engine_version: int
+    :param token: Authentication token to include in requests sent to Vault.
+        (for ``token`` and ``github`` auth_type)
+    :type token: str
+    :param role_id: Role ID for Authentication (for ``approle`` auth_type)
+    :type role_id: str
+    :param kubernetes_role: Role for Authentication (for ``kubernetes`` auth_type)
+    :type kubernetes_role: str
+    :param kubernetes_jwt_path: Path for kubernetes jwt token (for ``kubernetes`` auth_type, default:
+        ``/var/run/secrets/kubernetes.io/serviceaccount/token``)
+    :type kubernetes_jwt_path: str
+    :param gcp_key_path: Path to GCP Credential JSON file (for ``gcp`` auth_type)
+    :type gcp_key_path: str
+    :param gcp_scopes: Comma-separated string containing GCP scopes (for ``gcp`` auth_type)
+    :type gcp_scopes: str
+
+    """
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        vault_conn_id: str,
+        auth_type: Optional[str] = None,
+        mount_point: Optional[str] = None,
+        kv_engine_version: Optional[int] = None,
+        token: Optional[str] = None,
+        role_id: Optional[str] = None,
+        kubernetes_role: Optional[str] = None,
+        kubernetes_jwt_path: Optional[str] = None,
+        gcp_key_path: Optional[str] = None,
+        gcp_scopes: Optional[str] = None,
+    ):
+        super().__init__()
+        self.connection = self.get_connection(vault_conn_id)
+
+        if not auth_type:
+            auth_type = self.connection.extra_dejson.get('auth_type') or "token"
+
+        if not mount_point:
+            mount_point = self.connection.extra_dejson.get('mount_point')
+        if not mount_point:
+            mount_point = 'secret'
+
+        if not kv_engine_version:
+            conn_version = self.connection.extra_dejson.get("kv_engine_version")
+            try:
+                kv_engine_version = int(conn_version) if conn_version else DEFAULT_KV_ENGINE_VERSION
+            except ValueError:
+                raise VaultError(f"The version is not an int: {conn_version}. ")
+
+        if auth_type in ["approle"]:
+            if not role_id:
+                role_id = self.connection.extra_dejson.get('role_id')
+
+        if auth_type in ['github', "token"]:
+            if not token:
+                token = self.connection.extra_dejson.get('token')
+        gcp_key_path, gcp_scopes = \
+            self._get_gcp_parameters_from_connection(gcp_key_path, gcp_scopes) \
```

Comment:
Why do we pass `gcp_key_path, gcp_scopes` in to this function (same for kube below) -- aren't they guaranteed to only ever by None as they are local variables that are assigned a value based on the return

Reply:
Nope. They are parameter of constructor. Can be overridden.

Reply:
Do we need/want them to be in extra and constructor args?

Reply:
We do not need to. But if found it useful to be able to override some of the parameters - especially that this gives indication of what extras you can provide in the vault connection:

```
    The extras in the connection are named the same as the parameters (`mount_point`,'kv_engine_version' ...).
```

Since this is not basic authentication information (host/user/password/port which are part of the connection) but they are put in the extras. My assumption was that any "extra" can be also added as initialisation parameter. I looked at other hooks and we do it differently in different hooks, but usually when there are "extras", you can set those extras  in the hook init param (also because setting extras via connection is not straightforward if you do not add your own connection type (and then it's not backportable). So this is mostly convenience also useful when you start playing with the hook and would like to try it before having to put them in connection extras.

But I have no strong opinion here. 

I am fine to remove those if you think it's not good. Particularly in the next PR i will definitely remove keyfile_dict (as it contains secret information that should not be put as parameter). But gcp_key_path and gcp_scopes fall IMHO into  the same category as all other extras (kubernetes_role, kv_engine_version etc.) so we either keep them all (mostly for  convenience) or remove them all. 

I am fine with both -  if you have strong opinion I am happy to follow it.